ztools/telnet/telnet.go

```diff
@@ -0,0 +1,137 @@
+/*
+ * ZGrab Copyright 2015 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package telnet
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net"
+)
+
+// RFC 854 - https://tools.ietf.org/html/rfc854
+const (
+	IAC                = byte(255) //Interpret as command
+	DONT               = byte(254)
+	DO                 = byte(253)
+	WONT               = byte(252)
+	WILL               = byte(251)
+	GO_AHEAD           = byte(249) // Special go ahead command
+	IAC_CMD_LENGTH     = 3         // IAC commands take 3 bytes (inclusive)
+	READ_BUFFER_LENGTH = 8192
+)
+
+func GetTelnetBanner(logStruct *TelnetLog, conn net.Conn, maxReadSize int) (err error) {
+	if err = NegotiateOptions(logStruct, conn); err != nil {
+		return err
+	}
+
+	//grab banner
+	buffer := make([]byte, READ_BUFFER_LENGTH)
+
+	numBytes := len(buffer)
+	rounds := int(math.Ceil(float64(maxReadSize) / READ_BUFFER_LENGTH))
+	count := 0
+	for numBytes != 0 && count < rounds && numBytes == READ_BUFFER_LENGTH {
```

Comment:
Ah, I did reverse the meaning of those variables. In that case we should use `append()` to join the buffers populated by `conn.Read()`, and then convert to a string outside of the loop.